### PR TITLE
[3.0 Cleanup] Removal of the SecurityContext 

### PR DESCRIPTION
**Pull request by linaori**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | yes |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | ~ |

As follow up for PR https://github.com/symfony/symfony/pull/11690 in 2.6, I have removed the `SecurityContext`, `SecurityContextInterface` and `security.context` service. I have split the parts in separate commits to make it easier to review but I can squash them if desired.

~~I will submit the changes to the documentation when a branch is available to do so.~~ Docs are already updated.

### CODE REVIEW

**Comment by fabpot** [comment]
@iltar We are now ready to merge such PRs. Can you rebase on current master? Thanks.

**Comment by iltar** [comment]
@fabpot I will start working on it

**Comment by fabpot** [comment]
Thanks.

**Comment by iltar** [comment]
Seems like there are some unrelated test broken in the `FrameworkBundle`. As far as I can see, the tests and code I re-based is still working.

File: UPGRADE-3.0.md [hunk 1]
```diff
@@ -511,6 +512,27 @@ UPGRADE FROM 2.x to 3.0
 ### Security
 
  * The `Resources/` directory was moved to `Core/Resources/`
+ * [BC BREAK] All usages of `SecurityContext` and `SecurityContextInterface` have been replaced by either the
+   `TokenStorageInterface` or `AuthorizationCheckerInterface`, this includes _all_ class.
```
Comment (nicoschoenmaker) [inline UPGRADE-3.0.md]: Maybe change to "this includes _all_ classes"?

**Comment by fabpot** [comment]
@iltar Can you rebase so that we can merge it? Thanks.

**Comment by iltar** [comment]
@fabpot The tests are flaky. Sometimes it runs out of memory, sometimes I get failing tests and sometimes they all succeed. I think it might have been cache related.

**Comment by fabpot** [comment]
@iltar Sorry to ask you this again but can you rebase? Ping me when it's done and I will merge right away.

**Comment by iltar** [comment]
@fabpot rebase is done, I have no idea what phpunit did that made it unable to resolve a 2.8 dependency. I can't find the command that was executed that made fail so I cannot reproduce it myself, any way to find out which that was?

_No idea why 3.0 is tested against a 2.8 dependency_

**Comment by Tobion** [comment]
@nicolas-grekas should know about this (https://travis-ci.org/symfony/symfony/jobs/80268574#L2143)

File: UPGRADE-3.0.md [hunk 2]
```diff
@@ -607,6 +608,27 @@ UPGRADE FROM 2.x to 3.0
...
+ * [BC BREAK] Removed the `SecurityContext` class and `SecurityContextInterface`
+ * [BC BREAK] Constants from `SecurityContextInterface` are removed in favor of the constants in `Security`
```
Comment (sstok) [inline UPGRADE-3.0.md]: Maybe provide a Fully qualified class-name for `Security`?

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,10 @@
 CHANGELOG
 =========
 
+3.0.0
+-----
+ * `app.security` has been removed from twig
```
Comment (sstok) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: Should an empty line between header bullet-list.

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/CsrfFormLoginBundle/Controller/LoginController.php [hunk 4]
```diff
@@ -21,9 +21,9 @@ public function loginAction()
     {
         $form = $this->container->get('form.factory')->create('Symfony\Bundle\SecurityBundle\Tests\Functional\Bundle\CsrfFormLoginBundle\Form\UserLoginType');
 
-        return $this->container->get('templating')->renderResponse('CsrfFormLoginBundle:Login:login.html.twig', array(
+        return $this->container->get('templating')->renderResponse('CsrfFormLoginBundle:Login:login.html.twig', [
```
Comment (sstok) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/CsrfFormLoginBundle/Controller/LoginController.php]: I know @fabpot wanted to prevent these kind of changes as they may cause merge conflicts, and it looks inconsistent.

Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/CsrfFormLoginBundle/Controller/LoginController.php]: I agree, this should be reverted

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/FormLoginBundle/Controller/LocalizedController.php [hunk 5]
```diff
@@ -11,10 +11,10 @@
 
 namespace Symfony\Bundle\SecurityBundle\Tests\Functional\Bundle\FormLoginBundle\Controller;
 
-use Symfony\Component\Security\Core\Security;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\DependencyInjection\ContainerAware;
+use Symfony\Component\Security\Core\Security;
```
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/FormLoginBundle/Controller/LocalizedController.php]: reordering of the use statements must be reverted as well as it makes merging between branches more complex for no benefits.

**Comment by nicolas-grekas** [comment]
@iltar changing the composer.json **in 2.8** will fix the issue:
```diff
-        "symfony/http-kernel": "~2.2|~3.0.0"
+        "symfony/http-kernel": "~2.8"
```
this constraint should already come from framework-bundle since this is how it is there, but I guess composer has a bug here (mixing constraints from dev+no-dev+custom repository?)

File: UPGRADE-3.0.md (hunk 1 above)
Comment (stof) [inline UPGRADE-3.0.md]: What does ``this includes all class`` mean ?

File: src/Symfony/Bridge/Twig/CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,10 @@
 CHANGELOG
 =========
 
+3.0.0
+-----
+ * [BC BREAK] `SecurityExtension::__construct()` now uses `AuthorizationCheckerInterface` instead of `SecurityContextInterface`
```
Comment (stof) [inline src/Symfony/Bridge/Twig/CHANGELOG.md]: 2.8 already allows both though

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php [hunk 7]
```diff
@@ -11,12 +11,12 @@
...
 namespace Symfony\Bundle\SecurityBundle\DataCollector;
 
-use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
-use Symfony\Component\Security\Core\Role\RoleHierarchyInterface;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\HttpKernel\DataCollector\DataCollector;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
 use Symfony\Component\Security\Core\Role\RoleInterface;
+use Symfony\Component\Security\Core\Role\RoleHierarchyInterface;
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: reordering should be reverted

File: src/Symfony/Bundle/SecurityBundle/Tests/Templating/Helper/SecurityHelperTest.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+namespace Symfony\SecurityBundle\Tests\Templating\Helper;
+
+use Symfony\Bridge\Twig\Extension\SecurityExtension;
+use Symfony\Component\Security\Acl\Voter\FieldVote;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+
+/**
+ * @author Iltar van der Berg <ivanderberg@hostnet.nl>
+ */
+class SecurityHelperTest extends \PHPUnit_Framework_TestCase
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Tests/Templating/Helper/SecurityHelperTest.php]: adding new tests in a PR removing deprecated APIs is wrong. Such PR should only be removing code. Missing tests should be added in the older branch where they are worth it.

Btw, the name and namespace of the test is wrong. You are testing ``Symfony\Bridge\Twig\Extension\SecurityExtension``, so this test should be in the bridge and be named ``SecurityExtensionTest``

File: src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php [hunk 9]
```diff
@@ -44,7 +44,7 @@
     public function __construct(UserCheckerInterface $userChecker, $providerKey, $hideUserNotFoundExceptions = true)
     {
         if (empty($providerKey)) {
-            throw new \InvalidArgumentException('$providerKey must not be empty.');
+            throw new \InvalidArgumentException('$providerKey must not be empty. Did you forget to configure this?');
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php]: Improving exception messages should be done in older branches too

File: src/Symfony/Component/Security/Core/Tests/Validator/Constraints/UserPasswordValidatorTest.php [hunk 10]
```diff
@@ -79,10 +83,11 @@ public function testPasswordIsNotValid()
             'message' => 'myMessage',
         ));
 
-        $this->encoder->expects($this->once())
+        $this->encoder
+            ->expects($this->once())
             ->method('isPasswordValid')
             ->with(static::PASSWORD, 'secret', static::SALT)
-            ->will($this->returnValue(false));
+            ->willReturn(false);
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Tests/Validator/Constraints/UserPasswordValidatorTest.php]: switching to the new PHPUnit APIs should not be done in PRs cleaning deprecated layers. It is totally unrelated so it should be its own PR (and if we decide to do it, we should apply it in the 2.3 branch and then merge it to newer branches, otherwise it will become a huge pain for maintainers to merge branches together for the next 3 years)

File: src/Symfony/Component/Security/Core/Tests/Validator/Constraints/UserPasswordValidatorTest.php [hunk 11]
```diff
@@ -106,63 +111,53 @@ public function testUserIsNotValid()
...
-        ;
+            ->willReturn(static::SALT);
 
         return $mock;
     }
 
     protected function createPasswordEncoder($isPasswordValid = true)
     {
-        return $this->getMock('Symfony\Component\Security\Core\Encoder\PasswordEncoderInterface');
+        return $this->getMock(PasswordEncoderInterface::class);
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Tests/Validator/Constraints/UserPasswordValidatorTest.php]: this should be reverted. It makes merging branches harder (2.8 will keep using the string for the next 3 year of maintenance) and it does not add any real value to counterpart the maintenance pain

**Comment by stof** [comment]
I think it would be much easier to recreate the PR from scratch: 90% of the changes done in this PR have to be reverted.

**Comment by iltar** [comment]
@stof I agree, it's becoming a mess because the PR is pretty old. I will get to this the start of next week.
